pkg/interop/server_authn.go

```
@@ -88,102 +66,167 @@ nextPattern:
+}
+
+// authenticateNS authenticates the client as a Network Server.
+//
+// If the client presents a TLS client certificate, it is verified against the trusted CAs of the NetID.
+// Any DNS names in the X.509 Subject Alternative Names are taken as address patterns used to verify the NSID. If there
+// are no DNS names, the Common Name is used as the single address pattern.
+// If the TLS client certificate verification fails, this method returns an error.
+//
+// If the client presents a Bearer token in the Authorization header of the HTTP request, it is verified as token issued
+// by Packet Broker.
+func (s *Server) authenticateNS(ctx context.Context, r *http.Request, data []byte) (context.Context, error) {
+	var header NsMessageHeader
+	if err := json.Unmarshal(data, &header); err != nil {
+		return nil, ErrMalformedMessage.WithCause(err)
+	}
+	if !header.ProtocolVersion.SupportsNSID() && header.SenderNSID != nil {
+		return nil, ErrMalformedMessage.New()
+	}
+
+	for _, authFunc := range []func(context.Context) (*NetworkServerAuthInfo, error){
+		// Verify TLS client certificate.
+		func(ctx context.Context) (*NetworkServerAuthInfo, error) {
+			if r.TLS == nil || len(r.TLS.PeerCertificates) == 0 {
+				return nil, nil
+			}
+			addrs, err := s.verifySenderCertificate(ctx, types.NetID(header.SenderID).String(), r.TLS)
+			if err != nil {
+				return nil, err
+			}
+			return &NetworkServerAuthInfo{
+				NetID:     types.NetID(header.SenderID),
+				Addresses: addrs,
+			}, nil
+		},
+		// Verify token in a best-effort manner.
+		func(ctx context.Context) (*NetworkServerAuthInfo, error) {
+			logger := log.FromContext(ctx).WithField("authenticator", "packetbroker")
+			authz := strings.SplitN(r.Header.Get("Authorization"), " ", 2)
+			if len(authz) < 2 || strings.ToLower(authz[0]) != "bearer" {
+				return nil, nil
+			}
+			token, err := jwt.ParseSigned(authz[1])
+			if err != nil {
+				logger.WithError(err).Debug("Failed to parse token")
+				return nil, nil
```

This is pretty bad remote developer UX, but I guess we're swallowing the error such that we allow other authorization strategies to still run.

Another possibility is to aggregate the errors and to provide them as causes, and only return the aggregated error if none of the strategies match. If the error details are not rendered just skip this comment.

https://github.com/TheThingsNetwork/lorawan-stack/blob/658c8cef46fb7c397c19ee8f12446502fe8b68b3/pkg/applicationserver/io/pubsub/provider/connection.go#L98-L124

---

Hmm it is really best effort. We may add more token verifications and we would be trying them in no particular order.

The result is the same anyway: `UnknownSender` with status code 200.

The only difference is the logging really. Maybe we should warn here if the issuer matches but the verification doesn't?

---

But if we support API keys or other token formats in the future, we don't want previous verification attempts to always warn, right?

---

> The result is the same anyway: UnknownSender with status code 200. 

I was afraid this was the case. Then it doesn't help if we aggregate the errors or something, since the caller cannot hear the reasoning anyway.

> The only difference is the logging really. Maybe we should warn here if the issuer matches but the verification doesn't?

Not needed given that indeed we'll `WARN` even when another authorization strategy matches.